src/main/java/ArrayIndexedBag.java

```diff
@@ -0,0 +1,176 @@
+    private void shiftLeft(int start) {
+        for (int i = start; i < bag.length - 1; ++i) {
+            bag[i] = bag[i + 1];
+        }
+        bag[bag.length - 1] = null;
+    }
+
+    /**
+     * Shifts elements in an array up (towards index rear) away
+     * from the starting index specified. This method assumes that
+     * the calling function manages the rear field and any
+     * need of expandCapacity().
+     *
+     * @param start Index of where shifting starts/stops and where
+     *              space is opened.
+     */
+    private void shiftRight(int start) {
+        for (int i = rear; i > start; --i) {
+            bag[i] = bag[i - 1];
+        }
+        bag[start] = null;
+    }
+
+    @Override
+    public void add(T element) {
+        add(rear, element);
+    }
+
+    @Override
+    public void add(int index, T element) {
+        if (index > size()) {
+            throw new IndexOutOfBoundsException(String.format("Bag has no index %d to add to.", index));
+        }
+        if (size() == bag.length) {
+            expandCapacity();
+        }
+        shiftRight(index);
+        bag[index] = element;
+        rear++;
+    }
+
+    @Override
+    public void set(int index, T element) {
+        if (index >= size()) {
+            throw new IndexOutOfBoundsException(String.format("Bag has no element at index %d.", index));
+        }
+        bag[index] = element;
+    }
+
+    @Override
+    public T get(int index) {
+        if (index >= size()) {
+            throw new IndexOutOfBoundsException(String.format("Bag has no element at index %d.", index));
+        }
+        return bag[index];
+    }
+
+    @Override
+    public T remove(int index) {
+        if (index >= size()) {
+            throw new IndexOutOfBoundsException(String.format("Bag has no element at index %d.", index));
+        }
+        T returnElement = bag[index];
+        shiftLeft(index);
+        rear--;
+        return returnElement;
+    }
+
+    @Override
+    public int indexOf(T target) {
+        for (int i = 0; i < bag.length; ++i) {
+            if (bag[i].equals(target)) {
+                return i;
+            }
+        }
+        throw new NoSuchElementException("Element not contained in bag.");
+    }
+
+    @Override
+    public T remove(T element) {
+        if (isEmpty()) {
+            throw new NoSuchElementException("Removing from an empty bag.");
+        }
+        // If indexOf throws an exception, this method propagates it
+        int removeIndex = indexOf(element);
+        return remove(removeIndex);
+    }
+
+    @Override
+    public boolean contains(T target) {
+        try {
+            indexOf(target);
+            return true;
+        } catch (NoSuchElementException e) {
+            return false;
+        }
+    }
+
+    @Override
+    public int getCount(T target) {
+        int count = 0;
+        Iterator<T> it = this.iterator();
+        while (it.hasNext()) {
+            if (it.next().equals(target)) {
```

Comment:
Could do `interface Bag<T> extends Iterable<T>` so you can have `for (T t : this)` everywhere you use an iterator. Could also be a revisit, if you're doing iterators in more depth later.

Reply:
Perhaps I will:
0. Change to your succestion
1. In topic 15, show them the way it is without extends Iterable 
2. in topic 16 on iterators, I will show them this